client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/InvalidatedNearRemoteCache.java

```diff
@@ -0,0 +1,109 @@
+
+/**
+ * Near {@link org.infinispan.client.hotrod.RemoteCache} implementation
+ * enabling
+ *
+ * @param <K>
+ * @param <V>
+ */
+public class InvalidatedNearRemoteCache<K, V> extends RemoteCacheImpl<K, V> {
+
+   private final NearCacheService<K, V> nearcache;
+
+   public InvalidatedNearRemoteCache(RemoteCacheManager rcm, String name, NearCacheService<K, V> nearcache) {
+      super(rcm, name);
+      this.nearcache = nearcache;
+   }
+
+   @Override
+   public V get(Object key) {
+      VersionedValue<V> versioned = getVersioned((K) key);
+      return versioned != null ? versioned.getValue() : null;
+   }
+
+   @Override
+   public VersionedValue<V> getVersioned(K key) {
+      VersionedValue<V> nearValue = nearcache.get(key);
+      if (nearValue == null) {
+         VersionedValue<V> remoteValue = super.getVersioned(key);
+         if (remoteValue != null)
+            nearcache.putIfAbsent(key, remoteValue);
+
+         return remoteValue;
+      }
+
+      return nearValue;
+   }
+
+   @Override
+   public V put(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit) {
+      V ret = super.put(key, value, lifespan, lifespanUnit, maxIdleTime, maxIdleTimeUnit);
+      nearcache.remove(key); // Eager invalidation to avoid race
+      return ret;
+   }
+
+   @Override
+   public void putAll(Map<? extends K, ? extends V> map, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit) {
+      super.putAll(map, lifespan, lifespanUnit, maxIdleTime, maxIdleTimeUnit);
+      for (K k : map.keySet()) nearcache.remove(k);
+   }
+
+   @Override
+   public V replace(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit) {
+      V prev = super.replace(key, value, lifespan, lifespanUnit, maxIdleTime, maxIdleTimeUnit);
+      invalidateNearCacheIfNeeded(key, prev);
+      return prev;
+   }
+
+   @Override
+   public boolean replaceWithVersion(K key, V newValue, long version, int lifespanSeconds, int maxIdleTimeSeconds) {
+      boolean replaced = super.replaceWithVersion(key, newValue, version, lifespanSeconds, maxIdleTimeSeconds);    // TODO: Customise this generated block
+      if (replaced) nearcache.remove(key);
+      return replaced;
+   }
+
+   @Override
+   public V remove(Object key) {
+      V prev = super.remove(key);
+      invalidateNearCacheIfNeeded(key, prev);
+      return prev;
+   }
+
+   @Override
+   public boolean removeWithVersion(K key, long version) {
+      boolean removed = super.removeWithVersion(key, version);
+      if (removed) nearcache.remove(key); // Eager invalidation to avoid race
+      return removed;
+   }
+
+   @Override
+   public void clear() {
+      super.clear();
+      nearcache.clear(); // Clear near cache too
+   }
+
+   @SuppressWarnings("unchecked")
+   void invalidateNearCacheIfNeeded(Object key, Object prev) {
+      if (!operationsFactory.hasFlag(Flag.FORCE_RETURN_VALUE) || prev != null)
```

Comment:
It can't just be `prev != null` because a remove might return null even though it's removed things, but the only way you really know if the removed worked is based on parameter return. So, you either check the flag and then check the previous value, or you do it regardless of the return value...